pkg/agentscheduler/cache/shard_coordinator.go

```diff
@@ -0,0 +1,244 @@
+	cache                  Cache
+	updateChan             chan struct{} // channel for update notifications
+}
+
+type workerNodeShardState struct {
+	//revisionInScheduing revision of the NodeShard being used in worker scheduling. revisionInScheduing < 0 means no nodes in scheduling cycle
+	revisionInScheduing int64
+}
+
+func NewShardCoordinator(cache Cache, workerCount int, schedulerName string, shardingMode string) *ShardCoordinator {
+	klog.V(3).Infof("Shard Coordinator is initialized")
+	workerStates := make([]*workerNodeShardState, workerCount)
+	for i := range workerCount {
+		workerStates[i] = &workerNodeShardState{}
+	}
+
+	sc := &ShardCoordinator{
+		schedulerShardName: schedulerName,
+		workerStates:       workerStates,
+		shardingEnabled:    shardingMode == util.HardShardingMode || shardingMode == util.SoftShardingMode,
+		cache:              cache,
+		updateChan:         make(chan struct{}, 100), // buffered channel to handle multiple update requests
+	}
+	return sc
+}
+
+func (sc *ShardCoordinator) Run(stopCh <-chan struct{}) {
+	// Start the update processing goroutine
+	go sc.processUpdates(stopCh)
+}
+
+// getNodesForScheduling get nodes available to use for worker
+func (sc *ShardCoordinator) getNodesForScheduling(workerIdx int) sets.Set[string] {
+	if workerIdx >= len(sc.workerStates) {
+		klog.Errorf("Worker %d does not exist, no nodes are returned", workerIdx)
+		return sets.Set[string]{}
+	}
+	state := sc.workerStates[workerIdx]
+	if state == nil {
+		klog.Errorf("Worker %d state is not inited, no nodes are returned", workerIdx)
+		return sets.Set[string]{}
+	}
+	sc.mutex.RLock()
+	defer sc.mutex.RUnlock()
+	latest := atomic.LoadInt64(&sc.latestRevision)
+	atomic.StoreInt64(&state.revisionInScheduing, latest)
+	if klog.V(5).Enabled() {
+		klog.V(5).Infof("Worker %d will schedule with nodes%v", workerIdx, sc.nodeToUse.UnsortedList())
+	}
+	return sc.nodeToUse
+}
+
+// RefreshNodeShards update node shards cached in coordinator
+func (sc *ShardCoordinator) RefreshNodeShards(nodeShards map[string]*api.NodeShardInfo) {
+	if !sc.shardingEnabled {
+		return
+	}
+	sc.mutex.Lock()
+	sc.nodeShardInfos = nodeShards
+	shardForSchedulerFound := false
+	desiredNodesChanged := false
+	for shardName, shard := range nodeShards {
+		if shardName == sc.schedulerShardName {
+			if sc.schedulerNodeShardInfo == nil || !shard.NodeDesired.Equal(sc.schedulerNodeShardInfo.NodeDesired) {
+				desiredNodesChanged = true
+			}
+			sc.schedulerNodeShardInfo = shard
+			shardForSchedulerFound = true
+			break
+		}
+	}
+	needUpdate := false
+	if !shardForSchedulerFound {
+		klog.Errorf("Sharding is enabled but no shard is defined for this scheduler!")
+		sc.schedulerNodeShardInfo = nil
+	} else if availableNodes := sc.getAvailableNodesFromShard(); desiredNodesChanged || !sc.nodeToUse.Equal(availableNodes) {
+		atomic.AddInt64(&sc.latestRevision, 1)
+		sc.nodeToUse = availableNodes
+		needUpdate = true
+	}
+	sc.mutex.Unlock()
```

Comment:
The unlock is far from lock, which can be dangerous if someone return before Unlock